File: Q224 - Basic Calculator/0224.cpp
class Solution {
public:
    int calculate(string s) {

        int result=0, sumOrDiff=1;
        stack<int> stack;

        for(int i = 0; i < s.size(); i++) {
            char temp = s[i];
            if(temp >= '0' && temp <= '9') {
                int num = 0;
                while(i < s.size() && s[i]>='0' && s[i]<= '9')
                    num = num * 10 + (s[i++] - '0');
                result += sumOrDiff * num;
                i--;
            }
            else if(temp =='+') 
                sumOrDiff = 1;
            else if(temp =='-') 
                sumOrDiff = -1;
            else if(temp =='(') { 
                stack.push(result);
                stack.push(sumOrDiff);
                result = 0;
                sumOrDiff = 1;
            }
            else if(temp ==')') {
                result *= stack.top(); 
                stack.pop();
                result += stack.top(); 
                stack.pop();
            }
        }
        return result;
    }
};

File: Q-0224 - Basic Calculator/0224.py
class Solution:
    def calculate(self, s: str) -> int:
        
        result=0
        sumOrDiff=1
        stack=[]
        temp=0
        
        for char in s:
            if char.isdigit():
                temp=temp*10+int(char)
            elif char in ["-","+"]:
                result=result+temp*sumOrDiff
                temp=0
                if char=="-":
                    sumOrDiff=-1
                else:
                    sumOrDiff=1
            elif char=="(":
                stack.append(result)
                stack.append(sumOrDiff)
                sumOrDiff=1
                result=0
            elif char==")":
                result+=sumOrDiff*temp
                result*=stack.pop()
                result+=stack.pop()
                temp=0
        
        return result + temp * sumOrDiff
    